File: src/main/scala/com/lmmmowi/leetcode/scala/p34/Solution.scala
package com.lmmmowi.leetcode.scala.p34

/**
  * @Author: mowi
  * @Date: 2019-06-04
  * @Description: 34.在排序数组中查找元素的第一个和最后一个位置[https://leetcode-cn.com/problems/find-first-and-last-position-of-element-in-sorted-array/]
  */
object Solution {

    def searchRange(nums: Array[Int], target: Int): Array[Int] = {
        val left = search(nums, target, 0, nums.length - 1, true)
        val right = if (left == -1) -1 else search(nums, target, 0, nums.length - 1, false)
        Array[Int](left, right)
    }

    def search(nums: Array[Int], target: Int, start: Int, end: Int, findLeft: Boolean): Int = {
        if (start > end) {
            return -1
        }

        val mid = (start + end) / 2
        if (nums(mid) == target) {
            if (findLeft) {
                if (mid - 1 >= 0 && nums(mid - 1) == target) {
                    return search(nums, target, start, mid - 1, findLeft)
                }
                else {
                    return mid
                }
            }
            else {
                if (mid + 1 < nums.length && nums(mid + 1) == target) {
                    return search(nums, target, mid + 1, end, findLeft)
                }
                else {
                    return mid
                }
            }
        }
        else if (nums(mid) > target) {
            return search(nums, target, start, mid - 1, findLeft)
        }
        else {
            return search(nums, target, mid + 1, end, findLeft)
        }
    }

}

File: src/main/java/com/lmmmowi/leetcode/p34/Solution.java
package com.lmmmowi.leetcode.p34;

/**
 * @Author: mowi
 * @Date: 2020/6/19
 * @Description: 34.在排序数组中查找元素的第一个和最后一个位置[https://leetcode-cn.com/problems/find-first-and-last-position-of-element-in-sorted-array/]
 */
public class Solution {

    public static void main(String[] args) {
        int[] nums = new int[]{5, 7, 7, 8, 8, 10};
        int target = 6;
        int[] result = new Solution().searchRange(nums, target);
        System.out.println(result[0] + "," + result[1]);
    }

    private static final int LEFT = 0;
    private static final int RIGHT = 1;

    public int[] searchRange(int[] nums, int target) {
        int left = find(nums, 0, nums.length - 1, target, LEFT);
        int right = find(nums, 0, nums.length - 1, target, RIGHT);
        return new int[]{left, right};
    }

    private int find(int[] nums, int head, int tail, int target, int direction) {
        if (head > tail) {
            return -1;
        }

        int mid = (head + tail) / 2;

        if (nums[mid] > target) {
            return find(nums, head, mid - 1, target, direction);
        } else if (nums[mid] < target) {
            return find(nums, mid + 1, tail, target, direction);
        }

        if (LEFT == direction) {
            if (mid > 0 && nums[mid - 1] == target) {
                return find(nums, head, mid - 1, target, direction);
            }
        } else {
            if (mid + 1 < nums.length && nums[mid + 1] == target) {
                return find(nums, mid + 1, tail, target, direction);
            }
        }

        return mid;
    }
}